src/lib/incoming-channel.js

```diff
@@ -28,176 +28,121 @@ module.exports = class IncomingChannel {
+      throw new Error('incoming channel has not been created')
     }
 
-    if (channelId.toString('hex').toUpperCase() !== this._channelId) {
-      throw new Error('channel ID should be ' + this._channelId + '. Got: ' +
-        channelId.toString('hex').toUpperCase())
-    }
-
-*/
     const oldBalance = new BigNumber(yield this._balance.get())
+    debug('adding', transfer.amount, 'to', oldBalance.toString())
     const newBalance = oldBalance
       .add(transfer.amount)
-      .mul('1000000')
-/*
-    console.log('AMOUNT:', amount)
-    const signedAmount = new BigNumber(bignum.fromBuffer(amount, {
-      endian: 'big',
-      size: 8
-    }).toString())
-
-    if (!newBalance.eq(signedAmount)) {
-      throw new Error('claim amount doesn\'t match transfer amount. Transfer: '
-        + transfer.amount + ' XRP. Balance: '
-        + oldBalance.toString() + ' XRP. Claimed amount: '
-        + signedAmount.toString() + ' Drops.')
+
+    debug('processing claim for transfer with id:', transfer.id)
+    const ourClaim = encode.getClaimMessage(this._channelId, newBalance.toString())
+    debug('verifying signature "', claim, '" on', ourClaim.toString('hex'))
+    const verified = nacl.sign.detached.verify(
+      Buffer.from(ourClaim, 'hex'),
+      Buffer.from(claim, 'hex'),
+      this._publicKey)
+
+    if (!verified) {
+      // TODO: print the claim here
+      debug('signature was invalid')
+      throw new Error('signature (', claim, ') is invalid for claim:',
+        ourClaim)
     }
 
-    console.log('got claim for total ' +
-      signedAmount.toString() +
-      ' drops on transfer amount ' + transfer.amount)
-*/
+    debug('got valid claim for transfer id:', transfer.id,
+      'amount:', transfer.amount)
 
     // if the other side is sending claims we can't cash, then this will
     // figure it out
+    debug('adding transfer amount to balance')
     yield this._balance.add(transfer.amount)
     this._claim = claim
 
-    console.log('new balance:',
-      newBalance.toString(),
-      '\nmaximum:',
-      this._maximum.toString(),
-      '\nthreshold:',
-      this._maximum.mul(this._settlePercent).toString())
+    debug('checking threshold')
+    const threshold = this._balance.getMax().mul(this._settlePercent)
+    debug('new balance:', newBalance.toString(),
+      'maximum:', this._balance.toString(),
+      'threshold:', threshold.toString())
 
     // TODO: check if this should be claimed
-    if (newBalance.gt(this._maximum.mul(this._settlePercent).mul('1000000'))) {
+    if (newBalance.gt(threshold)) {
       // this should go asynchronously?
-      console.log('the new balance is greater than the threshold')
-      yield this._claimFunds()
-      //co(this._claimFunds.bind(this))
+      debug('new balance', newBalance.toString(),
+        'exceeds threshold', threshold.toString(),
+        '. submitting claim tx.')
+      yield co.wrap(this._claimFunds).call(this).catch((e) => {
```

Comment:
_claimFunds is idempotent, because XRP paychan lets you claim many times from a channel. Worst case is one of the claims could fail, but the other would still go through and so no money can be lost